app/lib/providers/authentication/authentication.dart

```diff
@@ -0,0 +1,58 @@
+import 'package:app/common/enums/app_state.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/material.dart';
+
+class AuthenticationProvider extends ChangeNotifier {
+  FirebaseAuth _firebaseAuth = FirebaseAuth.instance;
+  ViewState _viewState;
+
+  ViewState get viewState => _viewState;
+
+  void setViewState(ViewState viewState) {
+    _viewState = viewState;
+    notifyListeners();
+  }
+
+  Stream<User> get authStateChanges => _firebaseAuth.authStateChanges();
```

Comment:
I think I see what `StreamProvider` does here, providing two-way binding. 
- [ ] Why do we need both?

Reply:
The on the top is like initialization and this one is the implementation. Which will listen to auth changes like user sign in or sign out. This is the one that will decide which screen user should go to.